### PR TITLE
Remove MultiSend, move activeApps to FreeBalanceState, upgrade protocols, and more

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**
This pull request contains several changes rolled into one. They are:

### Removing `MultiSend`
Previously we used the `MultiSend` contract for the purposes of **atomically updating the free balance state and signing a multisignature wallet delegatecall transaction**. We needed this atomicity because we had a design goal of one-round-trip per protocol execution. We've relaxed this design goal and are okay with the protocols requiring multiple round trips, now.

### Add `activeApps` to the Free Balance state
Instead of using `uninstallKey` parameters on the pre-signed transaction made to an `AppInstance` to validate if it is in fact a funded App or not, we instead rely on a master list of "active" (a.k.a. funded) apps. This is an array on the state of the free balance. You can see how this logic plays out here:

https://github.com/counterfactual/monorepo/blob/ec25b2cb25942d1ed4e5eea8fe658f7ac652c30f/packages/contracts/contracts/ConditionalTransactionDelegateTarget.sol#L73-L86

This means the `UninstallKeyRegistry` is also removed.

### Updating protocols to manage effected changes

The following protocols have been upgraded:

- Install
- InstallVirtual
- Uninstall
- UninstallVirtual

In the case of the installation protocols, they now have an extra round trip per user involved. This is because to make an installation secure you need to _first_ see the unanimously signed conditional transaction (the call to the `ConditionalTransactionDelegateTarget`) and _then_ you feel OK with signing the update to the Free Balance (both `activeApps` and the relevant balance updates).

In the case of the uninstall, instead of setting a bit on the `UninstallKeyRegistry` now we set the state of the Free Balance object and remove the app from the `activeApps` list.

- [x] Still need to update the Mermaid diagrams!
- [x] Need to deploy new contract addresses

### Extend `FreeBalanceState`

https://github.com/counterfactual/monorepo/blob/ec25b2cb25942d1ed4e5eea8fe658f7ac652c30f/packages/node/src/models/free-balance.ts#L36-L41

Now it includes `balancesIndexedByToken` and `activeAppsBloomFilter` whereas previously it was just this:

```typescript
export type FreeBalanceState = {
  [tokenAddress: string]: PartyBalance[];
}
```

### Introduce `ConditionalTransaction`
This is a new commitment type which represents a multisignature wallet delegatecall function call to the `ConditionalTransactionDelegateTarget`. Previously this was encoded inside the "install commitment". See here:

Capture d’écran 2019-07-07 à 18 32 15

### Remove `TwoPartyVirtualEthAsLump.sol`

We had a special case contract that handled virtual app agreement interpretation in a hard-coded way for `TwoPartyFixedOutcome` outcome types. Instead, now we re-use `ConditionalTransactionDelegateTarget` as we do for all AppInstances but the clientside code is now hard-coded to the `TwoPartyFixedOutcomeETHInterpreter` for now instead.

### Introduce `IdentityApp`
For test scenarios this is a nice app. It is an app whose `computeOutcome` return value is its `encodedState`.

### Rename `TwoPartyEthAsLump ` to `TwoPartyFixedOutcomeETHInterpreter`
This name was also confusing and so now it references the exact outcome type and asset class it is meant to be an interpreter for. Additionally, the virtual equivalent `TwoPartyVirtualEthAsLump` which manages the distribution of assets between an endpoint and an intermediary was renamed to `TwoPartyFixedOutcomeFromVirtualAppETHInterpreter`.

### Rename `ETHBucket` to `FreeBalanceApp`
Since the Free Balance is taking on a larger role, I decided to give it a more unique name.

### Removing `rollup` from `@counterfactual/chain`
The only consumers of this package were internal to the codebase and didn't require anything other than access to the TypeScript source code, so I removed the rollup dependency and build step.

### Miscellaneous Code Cleanup
- Renamed `latestversionNumber` to `latestVersionNumber`
- Rename several function names to be more clear
- Make several methods in `free-balance.ts` functional instead of imperative
- Add lots of documentation in places
- Rename `validateSignature` to `requireValidSignatureOrThrowError`

### CODE REVIEW

**Comment by Alonski** [comment]
Maybe we should wait to merge in Joels fixes before merging this, or anything else, in?

**Comment by snario** [comment]
I'm going to split it up into multiple smaller PRs I think, consider this a draft atm

File: packages/node/src/models/free-balance.ts [hunk 1]
```diff
@@ -8,30 +7,44 @@ import { xkeysToSortedKthAddresses } from "../machine";
...
   amount: {
     _hex: string;
   };
-}
+};
 
-export interface FreeBalanceState {
-  [tokenAddress: string]: PartyBalance[];
-}
+export type ActiveAppsBloomFilter = { [appInstanceIdentityHash: string]: true };
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/models/free-balance.ts]: this is not a bloom filter...please don't call it one

File: packages/apps/migrations/2_deploy_contracts.js [hunk 2]
```diff
@@ -1,15 +1,10 @@
...
-const NimApp = artifacts.require("NimApp");
-const TicTacToeApp = artifacts.require("TicTacToeApp");
-const ETHUnidirectionalTransferApp = artifacts.require("ETHUnidirectionalTransferApp");
-
 const ARTIFACTS = [
-  HighRollerApp,
-  NimApp,
-  TicTacToeApp,
-  ETHUnidirectionalTransferApp
+  artifacts.require("HighRollerApp"),
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/apps/migrations/2_deploy_contracts.js]: separate PR

File: packages/apps/test/nim.spec.ts [hunk 3]
```diff
@@ -2,7 +2,6 @@ import { SolidityABIEncoderV2Type } from "@counterfactual/types";
 import chai from "chai";
 import * as waffle from "ethereum-waffle";
 import { Contract } from "ethers";
-import { AddressZero } from "ethers/constants";
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/apps/test/nim.spec.ts]: separate PR

File: packages/chain/package.json [hunk 4]
```diff
@@ -2,17 +2,14 @@
   "name": "@counterfactual/chain",
   "version": "0.0.1",
   "description": "Utility package to spin up a blockchain with deployed contracts for testing purposes",
-  "main": "dist/index.js",
-  "types": "dist/src/index.d.ts",
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/chain/package.json]: separate PR

File: packages/chain/src/contract-deployments.jest.ts [hunk 5]
```diff
@@ -1,28 +1,37 @@
...
-import TwoPartyEthAsLump from "@counterfactual/contracts/build/TwoPartyEthAsLump.json";
-import TwoPartyVirtualEthAsLump from "@counterfactual/contracts/build/TwoPartyVirtualEthAsLump.json";
-import UninstallKeyRegistry from "@counterfactual/contracts/build/UninstallKeyRegistry.json";
+import TwoPartyFixedOutcomeETHInterpreter from "@counterfactual/contracts/build/TwoPartyFixedOutcomeETHInterpreter.json";
+import TwoPartyFixedOutcomeFromVirtualAppETHInterpreter from "@counterfactual/contracts/build/TwoPartyFixedOutcomeFromVirtualAppETHInterpreter.json";
 import { NetworkContext } from "@counterfactual/types";
 import { ContractFactory, Wallet } from "ethers";
 
-export async function configureNetworkContext(wallet: Wallet) {
+export type NetworkContextForTestSuite = NetworkContext & {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/chain/src/contract-deployments.jest.ts]: separate PR

File: packages/contracts/contracts/interpreters/CoinTransferETHInterpreter.sol [hunk 6]
```diff
@@ -8,7 +8,7 @@ import "../libs/LibOutcome.sol";
  * This file is excluded from ethlint/solium because of this issue:
  * https://github.com/duaraghav8/Ethlint/issues/261
  */
-contract ETHInterpreter is Interpreter {
+contract CoinTransferETHInterpreter is Interpreter {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/contracts/interpreters/CoinTransferETHInterpreter.sol]: Should use a better name...suggest `ETHOfCoinTransferInterpreter`

File: packages/contracts/contracts/interpreters/TwoPartyFixedOutcomeETHInterpreter.sol [hunk 7]
```diff
@@ -5,7 +5,7 @@ import "../interfaces/Interpreter.sol";
 import "../libs/LibOutcome.sol";
 
 
-contract TwoPartyEthAsLump is Interpreter {
+contract TwoPartyFixedOutcomeETHInterpreter is Interpreter {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/contracts/interpreters/TwoPartyFixedOutcomeETHInterpreter.sol]: why the name change? seems slightly worse IMO

File: packages/chain/src/contract-deployments.jest.ts [hunk 8]
```diff
@@ -1,28 +1,37 @@
...
+  TicTacToeApp: string;
+};
+
+export async function deployTestArtifactsToChain(wallet: Wallet) {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/chain/src/contract-deployments.jest.ts]: this rename can be a separate PR

File: packages/contracts/contracts/default-apps/IdentityApp.sol [hunk 9]
```diff
@@ -0,0 +1,17 @@
+pragma solidity 0.5.10;
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/contracts/default-apps/IdentityApp.sol]: can be added in a separate PR

File: packages/contracts/contracts/mixins/MixinChallengeRegistryCore.sol [hunk 10]
```diff
@@ -22,6 +22,14 @@ contract MixinChallengeRegistryCore is MChallengeRegistryCore {
     return appChallenges[identityHash];
   }
 
+  function getAppStateHash(bytes32 identityHash)
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/contracts/mixins/MixinChallengeRegistryCore.sol]: this function seems unused

File: packages/contracts/contracts/default-apps/FreeBalanceApp.sol [hunk 11]
```diff
@@ -5,22 +5,23 @@ import "../interfaces/CounterfactualApp.sol";
 import "../libs/LibOutcome.sol";
 
 
-contract ETHBucket is CounterfactualApp {
+contract FreeBalanceApp is CounterfactualApp {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/contracts/default-apps/FreeBalanceApp.sol]: Should be called "LedgerApp" or similar since it now controls both the unassigned balance and the list of active apps

File: packages/contracts/migrations/2_deploy_contracts.js [hunk 12]
```diff
@@ -1,29 +1,16 @@
...
-  ETHBucket,
-  MinimumViableMultisig,
-  MultiSend,
-  UninstallKeyRegistry,
-  ProxyFactory,
-  ConditionalTransactionDelegateTarget,
-  TwoPartyVirtualEthAsLump,
-  ETHInterpreter,
-  TwoPartyEthAsLump
+  artifacts.require("ChallengeRegistry"),
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/migrations/2_deploy_contracts.js]: separate PR

File: packages/contracts/test/challenge-registry-dispute.spec.ts [hunk 13]
```diff
@@ -66,7 +66,7 @@ describe("ChallengeRegistry Challenge", () => {
 
   let setStateAsOwner: (versionNumber: number, appState?: string) => Promise<void>;
   let latestState: () => Promise<string>;
-  let latestversionNumber: () => Promise<number>;
+  let latestVersionNumber: () => Promise<number>;
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/test/challenge-registry-dispute.spec.ts]: separate PR

File: packages/contracts/test/challenge-registry.spec.ts [hunk 14]
```diff
@@ -40,7 +40,7 @@ describe("ChallengeRegistry", () => {
   let cancelChallenge: () => Promise<void>;
   let sendSignedFinalizationToChain: () => Promise<any>;
   let latestAppState: () => Promise<string>;
-  let latestversionNumber: () => Promise<number>;
+  let latestVersionNumber: () => Promise<number>;
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/contracts/test/challenge-registry.spec.ts]: separate PR

File: packages/node/src/ethereum/virtual-app-set-state-commitment.ts [hunk 15]
```diff
@@ -64,7 +64,7 @@ export class VirtualAppSetStateCommitment extends EthereumCommitment {
   }
 
   // overrides EthereumCommitment::Transaction
-  public transaction(
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/ethereum/virtual-app-set-state-commitment.ts]: separate PR

File: packages/node/src/methods/app-instance/propose-install-virtual/operation.ts [hunk 16]
```diff
@@ -68,47 +68,48 @@ export function getNextNodeAddress(
...
+  thisAddress: string,
+  intermediaries: string[]
 ): boolean {
   return intermediaries.includes(thisAddress);
 }
 
-export async function getOrCreateVirtualChannel(
-  initiatorIdentifier: string,
-  respondingIdentifier: string,
+export async function getOrCreateStateChannelThatWrapsVirtualAppInstance(
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/methods/app-instance/propose-install-virtual/operation.ts]: separate PR

File: packages/node/src/methods/app-instance/propose-install-virtual/operation.ts [hunk 17]
```diff
@@ -68,47 +68,48 @@ export function getNextNodeAddress(
...
-  thisAddress: Address,
-  intermediaries: Address[]
+  thisAddress: string,
+  intermediaries: string[]
 ): boolean {
   return intermediaries.includes(thisAddress);
 }
 
-export async function getOrCreateVirtualChannel(
-  initiatorIdentifier: string,
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/methods/app-instance/propose-install-virtual/operation.ts]: good catch

File: packages/node/src/methods/app-instance/propose-install-virtual/operation.ts [hunk 18]
```diff
@@ -68,47 +68,48 @@ export function getNextNodeAddress(
...
+  initiatorXpub: string,
+  respondingXpub: string,
   intermediaries: string[],
   store: Store
 ): Promise<StateChannel> {
-  let channel: StateChannel;
+  let stateChannel: StateChannel;
+
   try {
-    channel = await getChannelFromPeerAddress(
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/node/src/methods/app-instance/propose-install-virtual/operation.ts]: separate PR

File: packages/contracts/contracts/interpreters/CoinTransferETHInterpreter.sol (hunk 6 above)
Comment (snario) [reply packages/contracts/contracts/interpreters/CoinTransferETHInterpreter.sol]: TBH I think `CoinTransferETHInterpreter` is a better name than `ETHOfCoinTransferInterpreter`

File: packages/contracts/contracts/interpreters/TwoPartyFixedOutcomeETHInterpreter.sol (hunk 7 above)
Comment (snario) [reply packages/contracts/contracts/interpreters/TwoPartyFixedOutcomeETHInterpreter.sol]: - `Lump` is never used in any other context
- `ETH` should always be capitalized
- The outcome type is `TwoPartyOutcome` not `TwoParty`
- Standardizes the interpreter contract name as `<OutcomeType><AssetType>Interpreter`

I think it is much better

File: packages/contracts/contracts/default-apps/FreeBalanceApp.sol (hunk 11 above)
Comment (snario) [reply packages/contracts/contracts/default-apps/FreeBalanceApp.sol]: I think FreeBalance is more specific than Ledger because there is no on-chain methods to withdraw from it like there is in Nitro, instead it represents general structural data on the StateChannel. Maybe something like `RootApp` or something like this is clearer but for now I think `FreeBalance` is OK

File: packages/node/test/machine/integration/install-then-set-state.spec.ts [hunk 19]
```diff
@@ -55,7 +58,7 @@ beforeAll(async () => {
 /**
  * @summary Set up a StateChannel and then install a new AppInstance into it.
  *
- * @description We re-use the ETHBucket App (which is the app ETH Free Balance uses)
+ * @description We re-use the FreeBalanceApp App (which is the app ETH Free Balance uses)
```
Comment (cf19drofxots) [inline packages/node/test/machine/integration/install-then-set-state.spec.ts]: What is `ETH Free Balance` in this case?

File: packages/node/test/machine/integration/install-then-set-state.spec.ts [hunk 20]
```diff
@@ -137,45 +144,68 @@ describe("Scenario: install AppInstance, set state, put on-chain", () => {
...
-        freeBalanceETH.timeout,
-        appInstance.appSeqNo,
-        network.ETHInterpreter,
+        identityAppInstance.identityHash,
+        stateChannel.freeBalance.identityHash,
+        network.CoinTransferETHInterpreter,
         defaultAbiCoder.encode(
-          ["uint256"],
-          [freeBalanceETH.coinTransferInterpreterParams!.limit]
+          ["tuple(uint256 limit)"],
```
Comment (cf19drofxots) [inline packages/node/test/machine/integration/install-then-set-state.spec.ts]: This encoding should be defined in one place and used throughout the rest of the codebase.

File: packages/specs/source/peer-protocol.md [hunk 21]
```diff
@@ -32,28 +30,21 @@ A protocol consists of the following components:
 
 ## Primitive Types
 
-|   Type    |          Description          |                                                      Sample                                                       |
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/specs/source/peer-protocol.md]: we are using this extension: https://marketplace.visualstudio.com/items?itemName=shuworks.vscode-table-formatter

please do not format it any other way

File: packages/simple-hub-server/src/middlewares/validate-signature.ts [hunk 22]
```diff
@@ -18,10 +18,10 @@ import { AuthenticatedRequest, JsonApiResource } from "../types";
  * `signature_required` error.
  */
 export default function(app: Application) {
-  return validateSignature(app);
+  return requireValidSignatureOrThrowError(app);
 }
 
-function validateSignature(app: Application) {
+function requireValidSignatureOrThrowError(app: Application) {
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline packages/simple-hub-server/src/middlewares/validate-signature.ts]: I prefer the more concise `assertIsValidSignature`
